drake/examples/RobustBalancing/LIPM_3D_nstep.m

```diff
@@ -1,112 +1,146 @@
 z_nom = 1; % nominal center of mass height
 step_max = .7; % max step distance
 
+%% Create SOS program
+prog = spotsosprog;
+
+%% Create indeterminates
 [prog,t]=prog.newIndeterminate('t',1);
 [prog,q]=prog.newIndeterminate('q',2);
 [prog,v]=prog.newIndeterminate('v',2);
-[prog,u]=prog.newIndeterminate('u',2);
+if n > 0
+  [prog,u]=prog.newIndeterminate('u',2);
+end
 x = [q;v];
 
-
+%% Create polynomials V(t,x) and W(x)
 if time_varying
   V_vars = [t;x];
 else
   V_vars = x;
 end
-
 W_vars = x;
-
 [prog,V] = prog.newFreePoly(monomials(V_vars,0:degree));
 [prog,W] = prog.newFreePoly(monomials(W_vars,0:degree));
 
+%% LIPM dynamics
 f = [v;q*g/z_nom];
 
-% time rescaling
+% Time rescaling
+% tau = t / T
+% dx/dtau = dx/dt * dt/dtau = dx/dt*T
 T_unscaled = T;
 f = f*T;
 T = 1;
 
 Vdot = diff(V,x)*f + diff(V,t);
 
+%% Goal region
+if n > 0
+  % jump equation
+  % control input changes q only
+  % qp = qm - u
+  xp = [q-u;v];
 
-% from Koolen et. al IJRR
-% regions should depend on the instantaneous capture point
-r_ic = q + v*sqrt(z_nom/g);
-%% jump equation
-% control input changes q only
-% qp = qm - u 
-xp = [q-u;v];
-
-% goal region is based off V from 0-step model
-% V0p(x) = V0(0,xp)
-V0p = subs(V0,[x;t],[xp;0]);
+  % for n > 0, goal region is based off V from 0-step model
+  % V0p(x) = V0(0,xp)
+  V0p = subs(V0,[x;t],[xp;0]);
+else
+  % use a small radius around the origin
+  goal_radius = .01; % TODO: setting
+  V0p = goal_radius^2 - x'*x;
+end
 
+% State constraint
+A = diag(1./(R_diag.^2));
+h_X = 1 - x'*A*x;
 
-% 4 SOS constraints
-% (1) V(T,x) >= 0 for x in goal region
-% (2) -Vdot(t,x) <= 0 for x in X
-% (3) W(x) >= 0 for x in X
-% (4) W(x) >= V(0,x) + 1 for x in X
-sos = [(subs(V,t,T))*(1+x'*x + u'*u);-Vdot; W;W - subs(V,t,0) - 1];
+%% SOS constraints
+sos = msspoly;
+if n > 0
+  % (1) V(T,x) >= 0 for x in goal region
+  % goal region
+  [prog, goal_sos] = spotless_add_sprocedure(prog, (subs(V,t,T))*(1+x'*x + u'*u), V0p,[W_vars;u],2);
 
-R_diag = [2 2 2 2];
-A = diag(1./(R_diag.^2));
+  % state constraint
+  [prog, goal_sos] = spotless_add_sprocedure(prog, goal_sos, h_X,[W_vars;u],degree);
 
-h_X = 1 - x'*A*x;
+  % control input limit -step_max <= u <= step_max
+  [prog, goal_sos] = spotless_add_sprocedure(prog, goal_sos, step_max^2-u'*u,[W_vars;u],degree);
+else
+  % (1) V(t,x) >= 0 for x in goal region
+  [prog, goal_sos] = spotless_add_sprocedure(prog, V, V0p,V_vars,degree-2);
 
-% add goal region and h_X constraints
-[prog, sos(1)] = spotless_add_sprocedure(prog, sos(1), V0p,[W_vars;u],2);
-[prog, sos(1)] = spotless_add_sprocedure(prog, sos(1), h_X,[W_vars;u],degree);
-[prog, sos(2)] = spotless_add_sprocedure(prog, sos(2), h_X,V_vars,degree-2);
-[prog, sos(3)] = spotless_add_sprocedure(prog, sos(3), h_X,W_vars,degree-2);
-[prog, sos(4)] = spotless_add_sprocedure(prog, sos(4), h_X,W_vars,degree-2);
+  if time_varying
+    [prog, goal_sos] = spotless_add_sprocedure(prog, goal_sos, T^2-t^2,V_vars,degree-2);
```

Comment:
I was wondering: why is this here? The step time shouldn't matter for 0-step capturability, right?